rclcpp/include/rclcpp/client.hpp

```diff
@@ -355,28 +543,123 @@ class Client : public ClientBase
       >::value
     >::type * = nullptr
   >
-  SharedFutureWithRequest
+  SharedFutureWithRequestAndRequestId
   async_send_request(SharedRequest request, CallbackT && cb)
   {
-    SharedPromiseWithRequest promise = std::make_shared<PromiseWithRequest>();
-    SharedFutureWithRequest future_with_request(promise->get_future());
+    auto & req = *request;
+    PromiseWithRequest promise;
+    auto shared_future = promise.get_future().share();
+    auto req_id = async_send_request_impl(
+      req,
```

Comment:
The reference doesn't seem necessary, why not just:

```suggestion
    PromiseWithRequest promise;
    auto shared_future = promise.get_future().share();
    auto req_id = async_send_request_impl(
      *request,
```

Reply:
It was necessary, but only because was tricky.
Fixed in https://github.com/ros2/rclcpp/pull/1734/commits/77f2be0c28690b63ec7e4ccc1736dc3282105fa0.